### ISSUE TITLE
Set vs Hashmap

### ISSUE THREAD

**Issue by bencoveney** [acutmore/proposal-keyby#7]
Excuse me sticking my nose in 😄 I was having a read through and wondered whether this would feel right as an extension to `Set`. For example, in the README currently:

> There is no capability to override this behavior and allow _two different objects_ to be treated equal within the collection.

This sounds at-odds to me with the basic description of a set.

> The Set object lets you store _unique values_ of any type...
>
> _From https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set_

The `keyBy` callback is effectively a hashing function, making Set behave like a Hashmap. For that reason, I would think alternative to this would be to add a `Hashmap` or `Hashset` collection instead, into which people could pass `JSON.stringify` or whatever hashing behaviour they would like.

To me this would feel like a better conceptual fit. If I received a Hashmap from some calling code, I think I would be aware of the fact that Hashmaps make less guarantees around uniquess and could proceed with caution. If I receive a `Set` on the other hand, I may be less likely to question the uniqueness check and may be bitten by `keyBy` being configured in a way I didn't expect. Another way to say it would be that `Set` becomes less "trustworthy".

The are some equivalents to this stuff in C#: [Hashset](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=net-8.0) and [Hashtable](https://learn.microsoft.com/en-us/dotnet/api/system.collections.hashtable?view=net-8.0). I bring them up as points of comparison, because:
- Their names include the word "hash" - giving you a clue that values may not be unique as a consumer.
- The hash behaviour has always been modifiable (citation needed), so they haven't broken assumptions by introducing it at a later date.

**Comment by ljharb** [comment]
The point is to change the definition of "unique". Set's existing definition doesn't actually match the language's already - `-0` and `-0` are treated the same.

**Comment by acutmore** [comment]
A significant part of me agrees that `HashMap` or `HashSet` might be good names for this; it's what other languages have. But I keep feeling like it's not quite right because:

- would developers take it to imply that regular `Set` and `Map` are not "hash maps" when under-the-hood they are.
- the `keyBy` function doesn't return a pigeon box hash, it doesn't reduce the domain size - it's more providing an "identity" but `IdentityMap` is definitely a bad name and `CustomIdentityMap` is a mouthful

**Comment by bencoveney** [comment]
Been a little while but re-reading my initial post now, I don't know if I feel that strongly about the naming argument.

The part which rings more true for me is that when working with a Set, either:

- I produced it myself (and know whether or not it has key mapping logic bundled inside), or
- I don't know whether it has key mapping logic inside (maybe because it came from a library I don't deeply understand, or it could be differently configured sets when passed at runtime) and I could potentially get a surprise when I interact with it.

```TypeScript
function doWorkWithSet(theSet) {
  theSet.has({ uuid: "ABCDE", id: 1 }) // ?
}

doWorkWithSet(new Set([{ uuid: "ABCDE", id: 1 }]))
doWorkWithSet(new Set([{ uuid: "ABCDE", id: 1 }], { keyBy: (v) => v.uuid }))
doWorkWithSet(new Set([{ uuid: "ABCDE", id: 1 }], { keyBy: (v) => v.id }))
```

Maybe passing `keyBy` as an option when interacting with the `Set` (rather than tying that logic "inside") would mitigate that, e.g.

```TypeScript
const keyBySet = new Set([]);
keyBySet.add({ uuid: "ABCDE", id: 1 }, { keyBy: (v) => v.uuid });
keyBySet.add({ uuid: "ABCDE", id: 2 }, { keyBy: (v) => v.uuid });
keyBySet.has({ uuid: "ABCDE", id: 3 }, { keyBy: (v) => v.uuid }); // true
[...keyBySet];
```

But then I'm wondering why use a Set for this, rather than a map with a `CompositeKey`. Isn't Map effectively a Set where you can control uniqueness, because the keys are within your control? But the fact uniqueness has changed is expressed more explicitly. The cost is that you'd need to be using the `keyBy` function on each insertion/retrieval from the Map, instead of only when constructing the Set - I don't know if that would be impractical.

> the keyBy function doesn't return a pigeon box hash, it doesn't reduce the domain size

Is this "true in principle, but not in practice"? Given the examples show N object instances mapping to M UUIDs, where N seems to be greater than M, given there are collisions.

I'm not sure I'm equipped with the knowledge to keep up with the discussions here but appreciate you humouring my questions anyway :D

**Comment by ljharb** [comment]
It's important that the configuration only be done at creation time, both so it can't be tampered with and so that every operation on the collection consistently uses the same operations.

If you're calling `.add` on a Set passed from somewhere else, you have no guarantees whatsoever about what it does - you'd have to `Set.prototype.add.call` to be able to have assumptions in the first place.

**Comment by acutmore** [comment]
> you'd have to Set.prototype.add.call to be able to have assumptions in the first place.

That is one issue with allowing Map/Set to be constructed with the custom keyBy. Code that is currently using `Map.prototype.has` as a receiver brand check would no longer be side effect free. They would need to move to something that wouldn't trigger the callback such as: `Object.getOwnPropertyDescriptor(Map.prototype, "size").get`

**Comment by acutmore** [comment]
> Is this "true in principle, but not in practice"? Given the examples show N object instances mapping to M UUIDs, where N seems to be greater than M, given there are collisions.

Yeah from the perspective that a typical hash value is 32bit and when running V8 with compressed pointers object addresses also fit in 30bits. So in that sense the domains are the same size. What I should have focused on is if collisions are allowed. With a "hash" collisions are allowed and handled by an equality predicate, "keyBy" on the other hand maps exactly to the set of desired keys with no collisions. This is what makes it "simpler" (no secondary equality predicate) but also more expensive to compute.
